README.md
# jQuery TextFill

This jQuery plugin resizes text to make it fit into a container. The font size
gets as big as possible.

* [Homepage][index].
* [Simple example][demo].
* [Unit tests (assure plugin correctness)][tests].

![logo](http://jquery-textfill.github.io/images/logo.png)

## Usage

- Remember to include _jQuery_ and _jQuery TextFill_:

```html
<script src="jquery.min.js"></script>
<script src="jquery.textfill.min.js"></script>
```

- Select which element you'll use. Make sure to:
  - Specify the parent's width and height
  - Put the text inside of a `<span/>` child by default (see _Options_ to change this)

```html
<div id='my-element' style='width:100px;height:50px;'>
  <span>The quick brown fox jumps over the lazy dog</span>
</div>
```

- Initialize _jQuery TextFill_

```js
$('#my-element').textfill({
    ...options...
});
```

## Options

Remember, **container** means the _parent_ element, while **child** is the
element that will _resize_. On the example above, the parent was the `div` and the
child was the `span`.

| Name              | Description | Default Value |
| ----------------- | ----------- | ------------- |
| `minFontPixels`   | Minimal font size (in pixels). The text will shrink up to this value. | 4 |
| `maxFontPixels`   | Maximum font size (in pixels). The text will stretch up to this value.. If it's a negative value (`size <= 0`), the text will stretch to as big as the container can accommodate. | 40 |
| `innerTag`        | The child element tag to resize. We select it by using `$(innerTag + ':visible:first', container)` | `span` |
| `widthOnly`       | Will only resize to the width restraint. The font might become tiny under small containers.  | `false` |
| `explicitWidth`   | Explicit width to resize. Defaults to the container's width. | `null` |
| `explicitHeight`  | Explicit height to resize. Defaults to the container's height. | `null` |
| `changeLineHeight`| Also change the `line-height` of the parent container. This might be useful when shrinking to a small container. | `false` |
| `allowOverflow`   | Allows text to overflow when minFontPixels is reached. Won't fail resizing, but instead will overflow container. | `false` |
| `debug`           | Output debugging messages to console. | `false` |

For example,

```html
<script>
$(function() {
    $('#my-element').textfill({
        maxFontPixels: 36
    });
});
</script>
```

## Callbacks

| Name       | Called when...                       | Default Value |
| ---------- | ------------------------------------ | ------------- |
| `success`  | Called when a resizing is successful | `null`        |
| `fail`     | Called when a resizing is failed     | `null`        |
| `complete` | Called when all elements are done    | `null`        |

For example,

```html
<script>
$(function() {
    $('#my-element').textfill({
        success: function() {
		    console.log("yay!")
		},
		fail: function() {
		    alert("boo hoo!")
		}
    });
});
</script>
```

## Contributing

You are very welcome to contribute!
[A good number of people did](https://github.com/jquery-textfill/jquery-textfill/graphs/contributors),
so feel free to help no matter how small the changes might be.

Just _make sure_ to read the file [`CONTRIBUTING.md`](CONTRIBUTING.md) first.
There we make a quick take on how you could help us.

Also, there we lay down our rules for _reporting issues_ and _making pull
requests_. Gotcha! Now you can't say we didn't tell you about it!

If you found something critical or just want to make a suggestion
[open an issue][issue] and start typing right away.

## Credits

This jQuery plugin was created by [Russ Painter][russ] around May 2009,
beginning with a StackOverflow [question][soq].

In very early 2012, [Yu-Jie Lin][yu] helped to move the project to GitHub with
version _0.1_ and obtained the clearly stated open source licensing from Russ.

Around July 2014 [Alexandre Dantas][alex] was made a contributor.

## License

`jquery-textfill` is licensed under the _MIT License_. See file
[`COPYING.md`](COPYING.md) to see what you can and cannot do with the source.

[index]:  http://jquery-textfill.github.io/
[demo]:   http://jquery-textfill.github.io/example/
[tests]:  http://jquery-textfill.github.io/unit-tests
[issue]:  https://github.com/jquery-textfill/jquery-textfill/issues
[soq]:    http://stackoverflow.com/questions/687998/auto-size-dynamic-text-to-fill-fixed-size-container
[russ]:   https://github.com/GeekyMonkey
[yu]:     https://github.com/livibetter
[alex]:   https://github.com/alexdantas



CHANGES.md
# CHANGES

## Future

Wait for fonts to load and re-size. Only available on some browsers.
https://stackoverflow.com/questions/5680013/how-to-be-notified-once-a-web-font-has-loaded
  document.fonts.ready.then(function () {
    alert('All fonts in use by visible text have loaded.');
      alert('Roboto loaded? ' + document.fonts.check('1em Roboto'));  // true
  });

Update size when page resizes for responsiveness. 

## Development

## Version 0.6.2 (2018-02-24)
* Performance improvement
* Added option to truncate without failing

## Version 0.6.1 (2018-02-23)

* Using the widthOnly option will apply a white-space:nowrap style so that the measured text won't wrap and therefore not have a proper width calculation. (by Klemens Ullman-Marx)
* Potential infinite loop prevented (by joel Anna)
* Code cleanups (by Klemens Ulman-Marx and GeekyMonkey)
* Distributable files are now in the /dist/ folder. Makefile updated.
* Remove deprecated "callback" callback
* Compatable with webpack module jquery reference
* add `allowOverflow` option (by lepolt)

## Version 0.6.0 (2014-08-19T21:12:07Z)

* add `changeLineHeight` (functionality to resize line-height of parent)
* greatly improved code readability and comments

## Version 0.5.0 (2014-07-17T15:37:15Z)

* made option `widthOnly` imply 'whitespace:nowrap'

## Version 0.4.0 (2013-08-16T17:17:02Z)

* add tests for callbacks
* add `fail` callback (#15)
* deprecate `callback` callback with `success` callback
* fix `widthOnly` option (#11, #4)

Height will not cause failure when `widthOnly` is set as long as width can fit into the container.

## Version 0.3.5 (2013-05-08T00:40:22Z)

* merge patch for compatibility to ZeptoJS (by sagens42 #13)

## Version 0.3.4 (2013-04-10T03:55:51Z)

* fix error when no console.debug.
* add test for debug flag.

## Version 0.3.3 (2013-03-25T23:35:21Z)

* retag for including the manefest in the tag.

## Version 0.3.2 (2013-02-09T02:43:32Z)

* fix introduced global variable fontSize by `_sizing()`
* fix container too small for bigger font size setting (#11)

## Version 0.3.1 (2013-01-27T05:10:51Z)

* Add `debug` option.
* Fix a couple of bugs, they didn't size up as big as possible

## Version 0.3 (2013-01-04T16:27:52Z)

* Allow to use *MaxFontPixels <= 0* to make sizing as big as possible.

## Version 0.2.1 (2012-11-11T07:04:20Z)

* Add explicitWidth and explicitHeight options (#6)

## Version 0.2 (2012-07-02T17:13:27Z)

* Fix resizing algorithm (#3)
* Add option widthOnly for header tags (#4)

## Version 0.1.3 (2012-03-27T04:03:30Z)

* Add callbacks for each filled and all filled. (by alex-pex #2)

## Version 0.1.2 (2012-01-29T12:46:12Z)

* Use binary search instead of plain do-while to accelerate the function.
  (by acsaga #1)

## Version 0.1.1

*  Add minFontPixels parameter

## 2012-01-16T16:56:32Z

*  This project was moved to GitHub and licensed under the MIT License.

## Version 0.1

*  Released in May 2009.



js/tests.js
var JTF = 'jtf';


function setup(opts) {
  var $t = $('#qunit-fixture');
  var $d = $('<div/>', opts.div).appendTo($t);
  var $s = $('<span/>', opts.span).appendTo($d);
}


test('capped at 10px', function () {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 10});
  equal($span.css('font-size'), '10px');
});


test('size up to max', function () {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 0});
  equal($span.css('font-size'), '172px');
});


test('width be maxWidth', function () {
  setup({
    div: {
      id: JTF,
      width: 196,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 0});
  equal($span.css('font-size'), '119px');
});


test('height be maxHeight', function () {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 158
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 0});
  equal($span.css('font-size'), '158px');
});


test('minFontPixels too big to fit in', function () {
  setup({
    div: {
      id: JTF,
      width: 40,
      height: 40
    },
    span: {
      css: {
        'font-family': 'VT323',
        'font-size': '20px',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, minFontPixels: 100, maxFontPixels: 0});
  equal($span.css('font-size'), '20px');
});


/*************/
/* callbacks */
/*************/

test('success callback', 1, function() {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    maxFontPixels: 0,
    success: function(e) {
      equal(e, $jtf[0]);
    }
  });
});


test('callback callback (deprecated)', 1, function() {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    maxFontPixels: 0,
    callback: function(e) {
      equal(e, $jtf[0]);
    }
  });
});


test('fail callback', 1, function () {
  setup({
    div: {
      id: JTF,
      width: 40,
      height: 40
    },
    span: {
      css: {
        'font-family': 'VT323',
        'font-size': '20px',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    minFontPixels: 100,
    maxFontPixels: 0,
    fail: function(e) {
      equal(e, $jtf[0]);
    }
  });
});


test('complete callback', 2, function() {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  $jtf.textfill({
    debug: true,
    maxFontPixels: 0,
    callback: function(e) {
      equal(e, $jtf[0]);
    },
    complete: function(e) {
      equal(e, $jtf);
    }
  });
});


/****************/
/* debug option */
/****************/

module('debug option', {
  setup: function () {
    if (console.debug_original) {
      throw 'console.debug_original already has value.';
    }
    console.debug_original = console.debug;
    console.debug_called = false;
    console.debug = function () {
      console.debug_called = true;
    }
  },
  teardown: function () {
    if (!console.debug_original) {
      throw 'console.debug_original is empty.';
    }
    console.debug = console.debug_original;
    console.debug_original = undefined;
    console.debug_called = undefined;
  }
});


test('debug used', function () {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  $jtf.textfill({debug: true, maxFontPixels: 10});
  equal(console.debug_called, true);
});


test('debug not used', function () {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  $jtf.textfill({debug: false, maxFontPixels: 10});
  equal(console.debug_called, false);
});




unit-tests/tests.js
/*global QUnit,$*/

/**
 * Unit Tests for jQuery TextFill.
 */

var JTF = 'jtf';

function setup(opts) {
  var $t = $('#qunit-fixture');
  var $d = $('<div/>', opts.div).appendTo($t);
  var $s = $('<span/>', opts.span).appendTo($d);
}

/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * Basic Tests
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

QUnit.test('capped at 10px', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 10});
  assert.equal($span.css('font-size'), '10px');
});


QUnit.test('size up to max', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 0});
  assert.equal($span.css('font-size'), '172px');
});


QUnit.test('width be maxWidth', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 196,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 0});
  assert.equal($span.css('font-size'), '119px');
});


QUnit.test('height be maxHeight', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 158
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, maxFontPixels: 0});
  assert.equal($span.css('font-size'), '158px');
});


QUnit.test('minFontPixels too big to fit in', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 40,
      height: 40
    },
    span: {
      css: {
        'font-family': 'VT323',
        'font-size': '20px',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({debug: true, minFontPixels: 100, maxFontPixels: 0});
  assert.equal($span.css('font-size'), '20px');
});


QUnit.test('minFontPixels too big to fit in, but widthOnly = True and width fits', function (assert) {
  // @ fontSize = 60 => H > 10, W = 100
  setup({
    div: {
      id: JTF,
      // just enough to fit in with minFontPixels
      width: 100,
      // H is busted, but widthOnly = True, so ignored
      height: 10
    },
    span: {
      css: {
        'font-family': 'VT323',
        'font-size': '20px',
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    minFontPixels: 20,
    maxFontPixels: 100,
    widthOnly: true
  });
  assert.equal($span.css('font-size'), '60px');
});


QUnit.test('minFontPixels too big to fit in, W/H both fail, even widthOnly = True', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 10,
      height: 10
    },
    span: {
      css: {
        'font-family': 'VT323',
        'font-size': '20px'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    minFontPixels: 20,
    maxFontPixels: 100,
    widthOnly: true
  });
  assert.equal($span.css('font-size'), '20px');
});


QUnit.test('allowOverflow will not result in fail callback', 1, function (assert) { 
  setup({
    div: {
      id: JTF,
      width: 60,
      height: 60
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'This text will cause overflow to occur, but will not result in a fail callback.'
    }
  });
  var $jtf = $('#' + JTF);
  $jtf.textfill({
    minFontPixels: 14,
    allowOverflow: true,
    fail: function(e) {
      assert.ok(false, 'fail callback should not have been called');
    },
    success: function (e) {
      assert.equal(e, $jtf[0]);
    }
  });
});

/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * Callbacks
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

QUnit.test('success callback', 1, function(assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    maxFontPixels: 0,
    success: function(e) {
      assert.equal(e, $jtf[0]);
    }
  });
});


QUnit.test('fail callback', 1, function (assert) {
  setup({
    div: {
      id: JTF,
      width: 40,
      height: 40
    },
    span: {
      css: {
        'font-family': 'VT323',
        'font-size': '20px'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  var $span = $jtf.find('span');
  $jtf.textfill({
    debug: true,
    minFontPixels: 100,
    maxFontPixels: 0,
    fail: function(e) {
      assert.equal(e, $jtf[0]);
    }
  });
});


QUnit.test('complete callback', 1, function(assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  $jtf.textfill({
    debug: true,
    maxFontPixels: 0,
    complete: function(e) {
      assert.equal(e, $jtf);
    }
  });
});


/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * "Debug" option
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

QUnit.module('debug option', {
  setup: function () {
    if (console.debug_original) {
      throw 'console.debug_original already has value.';
    }
    console.debug_original = console.debug;
    console.debug_called = false;
    console.debug = function () {
      console.debug_called = true;
    };
  },
  teardown: function () {
    if (!console.debug_original) {
      throw 'console.debug_original is empty.';
    }
    console.debug = console.debug_original;
    console.debug_original = undefined;
    console.debug_called = undefined;
  }
});


QUnit.test('debug used', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  $jtf.textfill({debug: true, maxFontPixels: 10});
  assert.equal(console.debug_called, true);
});


QUnit.test('debug not used', function (assert) {
  setup({
    div: {
      id: JTF,
      width: 285,
      height: 210
    },
    span: {
      css: {
        'font-family': 'VT323'
      },
      text: 'test'
    }
  });

  var $jtf = $('#' + JTF);
  $jtf.textfill({debug: false, maxFontPixels: 10});
  assert.equal(console.debug_called, false);
});


source/jquery.textfill.js
/**
 * @preserve  textfill
 * @name      jquery.textfill.js
 * @author    Russ Painter (GeekyMonkey)
 * @author    Yu-Jie Lin
 * @author    Alexandre Dantas
 * @version   0.6.2
 * @date      2018-02-24
 * @copyright (c) 2009
 * @license   MIT License
 * @homepage  https://github.com/jquery-textfill/jquery-textfill
 * @example   http://jquery-textfill.github.io/jquery-textfill/index.html
 */
; (function($) {

	/**
	 * Resizes an inner element's font so that the
	 * inner element completely fills the outer element.
	 *
	 * @param {Object} options User options that take
	 *                         higher precedence when
	 *                         merging with the default ones.
	 *
	 * @return All outer elements processed
	 */
	$.fn.textfill = function(options) {

		// ______  _______ _______ _______ _     _        _______ _______
		// |     \ |______ |______ |_____| |     | |         |    |______
		// |_____/ |______ |       |     | |_____| |_____    |    ______|
		//
		// Merging user options with the default values

		var defaults = {
			debug            : false,
			maxFontPixels    : 40,
			minFontPixels    : 4,
			innerTag         : 'span',
			widthOnly        : false,
			success          : null, // callback when a resizing is done
			fail             : null, // callback when a resizing is failed
			complete         : null, // callback when all is done
			explicitWidth    : null,
			explicitHeight   : null,
			changeLineHeight : false,
			truncateOnFail   : false,
			allowOverflow    : false // If true, text will stay at minFontPixels but overflow container w/out failing 
		};

		var Opts = $.extend(defaults, options);

		// _______ _     _ __   _ _______ _______ _____  _____  __   _ _______
		// |______ |     | | \  | |          |      |   |     | | \  | |______
		// |       |_____| |  \_| |_____     |    __|__ |_____| |  \_| ______|
		//
		// Predefining the awesomeness

		// Output arguments to the Debug console
		// if "Debug Mode" is enabled
		function _debug() {
			if (!Opts.debug
				||  typeof console       == 'undefined'
				||  typeof console.debug == 'undefined') {
				return;
			}
			console.debug.apply(console, arguments);
		}

		// Output arguments to the Warning console
		function _warn() {
			if (typeof console      == 'undefined' ||
				typeof console.warn == 'undefined') {
				return;
			}
			console.warn.apply(console, arguments);
		}

		// Outputs all information on the current sizing
		// of the font.
		function _debug_sizing(prefix, ourText, maxHeight, maxWidth, minFontPixels, maxFontPixels) {

			function _m(v1, v2) {

				var marker = ' / ';

				if (v1 > v2) {
					marker = ' > ';
				}
				else if (v1 == v2) {
					marker = ' = ';
				}
				return marker;
			}

			_debug(
				'[TextFill] '  + prefix + ' { ' +
				'font-size: ' + ourText.css('font-size') + ',' +
				'Height: '    + ourText.height() + 'px ' + _m(ourText.height(), maxHeight) + maxHeight + 'px,' +
				'Width: '     + ourText.width()  + _m(ourText.width() , maxWidth)  + maxWidth + ',' +
				'minFontPixels: ' + minFontPixels + 'px, ' +
				'maxFontPixels: ' + maxFontPixels + 'px }'
			);
		}

		/**
		 * Calculates which size the font can get resized,
		 * according to constrains.
		 *
		 * @param {String} prefix Gets shown on the console before
		 *                        all the arguments, if debug mode is on.
		 * @param {Object} ourText The DOM element to resize,
		 *                         that contains the text.
		 * @param {function} func Function called on `ourText` that's
		 *                        used to compare with `max`.
		 * @param {number} max Maximum value, that gets compared with
		 *                     `func` called on `ourText`.
		 * @param {number} minFontPixels Minimum value the font can
		 *                               get resized to (in pixels).
		 * @param {number} maxFontPixels Maximum value the font can
		 *                               get resized to (in pixels).
		 *
		 * @return Size (in pixels) that the font can be resized.
		 */
		function _sizing(prefix, ourText, func, max, maxHeight, maxWidth, minFontPixels, maxFontPixels) {

			_debug_sizing(
				prefix, ourText,
				maxHeight, maxWidth,
				minFontPixels, maxFontPixels
			);

			// The kernel of the whole plugin, take most attention
			// on this part.
			//
			// This is a loop that keeps increasing the `font-size`
			// until it fits the parent element.
			//
			// - Start from the minimal allowed value (`minFontPixels`)
			// - Guesses an average font size (in pixels) for the font,
			// - Resizes the text and sees if its size is within the
			//   boundaries (`minFontPixels` and `maxFontPixels`).
			//   - If so, keep guessing until we break.
			//   - If not, return the last calculated size.
			//
			// I understand this is not optimized and we should
			// consider implementing something akin to
			// Daniel Hoffmann's answer here:
			//
			//     http://stackoverflow.com/a/17433451/1094964
			//
			
			while (minFontPixels < (Math.floor(maxFontPixels) - 1)) {

				var fontSize = Math.floor((minFontPixels + maxFontPixels) / 2);
				ourText.css('font-size', fontSize);
				var curSize = func.call(ourText);

				if (curSize <= max) {
					minFontPixels = fontSize;

					if (curSize == max) {
						break;
					}
				}
				else {
					maxFontPixels = fontSize;
				}

				_debug_sizing(
					prefix, ourText,
					maxHeight, maxWidth,
					minFontPixels, maxFontPixels
				);
			}

			ourText.css('font-size', maxFontPixels);

			if (func.call(ourText) <= max) {
				minFontPixels = maxFontPixels;

				_debug_sizing(
					prefix + '* ', ourText,
					maxHeight, maxWidth,
					minFontPixels, maxFontPixels
				);
			}
			return minFontPixels;
		}

		// _______ _______ _______  ______ _______
		// |______    |    |_____| |_____/    |
		// ______|    |    |     | |    \_    |
        //
		// Let's get it started (yeah)!

		_debug('[TextFill] Start Debug');

		this.each(function() {

			// Contains the child element we will resize.
			// $(this) means the parent container
			var ourText = $(Opts.innerTag + ':first', this);

			_debug('[TextFill] Inner text: ' + ourText.text());
			_debug('[TextFill] All options: ', Opts);
			_debug('[TextFill] Maximum sizes: { ' +
				   'Height: ' + maxHeight + 'px, ' +
				   'Width: '  + maxWidth  + 'px' + ' }'
				  );

			if (!ourText.is(':visible')) {
				// Failure callback
				if (Opts.fail)
					Opts.fail(this);

				_debug(
					'[TextFill] Failure { ' +
					'Current Width: '  + ourText.width()  + ', ' +
					'Maximum Width: '  + maxWidth         + ', ' +
					'Current Height: ' + ourText.height() + ', ' +
					'Maximum Height: ' + maxHeight        + ' }'
				);

				return;
			}

			// Will resize to this dimensions.
			// Use explicit dimensions when specified
			var maxHeight = Opts.explicitHeight || $(this).height();
			var maxWidth  = Opts.explicitWidth  || $(this).width();

			var oldFontSize = ourText.css('font-size');

			var lineHeight  = parseFloat(ourText.css('line-height')) / parseFloat(oldFontSize);

			var minFontPixels = Opts.minFontPixels;

			// Remember, if this `maxFontPixels` is negative,
			// the text will resize to as long as the container
			// can accomodate
			var maxFontPixels = (Opts.maxFontPixels <= 0 ?
								 maxHeight :
								 Opts.maxFontPixels);

			// Let's start it all!

			// 1. Calculate which `font-size` would
			//    be best for the Height
			var fontSizeHeight = undefined;

			if (Opts.widthOnly) {
				// We need to measure with nowrap otherwise wrapping occurs and the measurement is wrong
      			ourText.css('white-space', 'nowrap' );
			} else {
				fontSizeHeight = _sizing(
					'Height', ourText,
					$.fn.height, maxHeight,
					maxHeight, maxWidth,
					minFontPixels, maxFontPixels
				);
			}

			// 2. Calculate which `font-size` would
			//    be best for the Width
			var fontSizeWidth = undefined;

			fontSizeWidth = _sizing(
				'Width', ourText,
				$.fn.width, maxWidth,
				maxHeight, maxWidth,
				minFontPixels, maxFontPixels
			);

			// 3. Actually resize the text!

			if (Opts.widthOnly) {
				ourText.css({
					'font-size'  : fontSizeWidth
				});

				if (Opts.changeLineHeight) {
					ourText.parent().css(
						'line-height',
						(lineHeight * fontSizeWidth + 'px')
					);
				}
			}
			else {
				var fontSizeFinal = Math.min(fontSizeHeight, fontSizeWidth);

				ourText.css('font-size', fontSizeFinal);

				if (Opts.changeLineHeight) {
					ourText.parent().css(
						'line-height',
						(lineHeight * fontSizeFinal) + 'px'
					);
				}
			}

			_debug(
				'[TextFill] Finished { ' +
				'Old font-size: ' + oldFontSize              + ', ' +
				'New font-size: ' + ourText.css('font-size') + ' }'
			);

			// Oops, something wrong happened!
			// If font-size increasing, we weren't supposed to exceed the original size 
			// If font-size decreasing, we hit minFontPixels, and still won't fit 
			if ((ourText.width()  > maxWidth && !Opts.allowOverflow) ||
				(ourText.height() > maxHeight && !Opts.widthOnly && !Opts.allowOverflow)) { 

				ourText.css('font-size', oldFontSize);

				// Failure callback
				if (Opts.fail) {
					Opts.fail(this);
				}

				_debug(
					'[TextFill] Failure { ' +
					'Current Width: '  + ourText.width()  + ', ' +
					'Maximum Width: '  + maxWidth         + ', ' +
					'Current Height: ' + ourText.height() + ', ' +
					'Maximum Height: ' + maxHeight        + ' }'
				);
			}
			else if (Opts.success) {
				Opts.success(this);
			}
		});

		// Complete callback
		if (Opts.complete) {
			Opts.complete(this);
		}

		_debug('[TextFill] End Debug');
		return this;
	};

})(function() {
	if (typeof module !== 'undefined' && module.exports) {
		return require('jquery');
	}
	return window.jQuery;
}());

